oxidation/libparsec/crates/core_fs/src/storage/local_database.rs

```diff
@@ -1,48 +1,150 @@
-// #9 Maximum Number Of Host Parameters In A Single SQL Statement
+use std::{
+    ops::DerefMut,
+    path::PathBuf,
+    sync::{Arc, Mutex, MutexGuard},
+};
+
+/// Maximum Number Of Host Parameters In A Single SQL Statement
+/// https://www.sqlite.org/limits.html#max_variable_number
 pub const SQLITE_MAX_VARIABLE_NUMBER: usize = 999;
-pub struct SqlitePool(Pool<ConnectionManager<SqliteConnection>>);
-pub type SqliteConn = PooledConnection<ConnectionManager<SqliteConnection>>;
 
-impl SqlitePool {
-    pub fn new<P: Into<String>>(path: P) -> FSResult<SqlitePool> {
+/// A Connection to Sqlite.
+#[derive(Clone)]
+pub struct SqliteConn {
+    connection: Arc<Mutex<Option<SqliteConnection>>>,
+}
+
+impl SqliteConn {
+    /// Create a new Sqlite connection with some initialization steps.
+    pub fn new<P: Into<String>>(path: P) -> FSResult<Self> {
         let path = path.into();
         if let Some(prefix) = PathBuf::from(&path).parent() {
             std::fs::create_dir_all(prefix).map_err(|_| FSError::CreateDir)?;
         }
-        let manager = ConnectionManager::<SqliteConnection>::new(path);
-        Pool::builder()
-            .build(manager)
-            .map_err(|_| FSError::Pool)
-            .map(Self)
-    }
-
-    pub fn conn(&self) -> FSResult<SqliteConn> {
-        let mut conn = self
-            .0
-            .get()
-            .map_err(|e| FSError::Connection(e.to_string()))?;
+
+        let mut connection =
+            SqliteConnection::establish(&path).map_err(|e| FSError::Connection(e.to_string()))?;
+
         // The combination of WAL journal mode and NORMAL synchronous mode
         // is a great combination: it allows for fast commits (~10 us compare
         // to 15 ms the default mode) but still protects the database against
         // corruption in the case of OS crash or power failure.
-        conn.batch_execute(
-            "
+        connection
+            .batch_execute(
+                "
             PRAGMA journal_mode = WAL; -- better write-concurrency
             PRAGMA synchronous = NORMAL; -- fsync only in critical moments
         ",
-        )
-        .map_err(|_| FSError::Configuration)?;
+            )
+            .map_err(|_| FSError::Configuration)?;
+
+        Ok(connection.into())
+    }
+
+    /// Get exclusive access to the Sqlite connection among cloned [SqliteConn]s.
+    pub fn lock_connection(&self) -> LockedSqliteConnection {
+        self.connection.lock().expect("Mutex is poisoned").into()
+    }
+
+    /// Close the connection to the local database.
+    pub fn close_connection(&self) {
+        *self.lock_connection().conn = None;
+    }
+
+    #[inline]
+    pub fn exec<T, F>(&self, f: F) -> Result<T, ExecError>
+    where
+        F: FnOnce(&mut SqliteConnection) -> Result<T, diesel::result::Error>,
+    {
+        self.lock_connection().exec(f)
+    }
+
+    #[inline]
+    pub fn exec_with_handler<T, F, G>(&self, f: F, error_handler: G) -> FSResult<T>
```

Comment:
Why do we need this, instead of `self.exec()?;` ?

Reply:
Because sometime we want to manually handle the database error, and we better refactor it latter